### DISCUSSION TITLE
.NET 9 Preview 2

### DISCUSSION THREAD

**Discussion by JonDouglas** [dotnet/core#9217] [News]
We are pleased to announce the release of .NET 9 Preview 2. 

This release contains the following Runtime & SDK improvements:
* RyuJIT enhancements
* Arm64 vectorization
* Parallel Testing & Terminal Logger Test Display

For details on these new features and more see the [Runtime Release Notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview2/runtime.md) and [SDK Release Notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview2/sdk.md)

You can find detailed release notes for additional features in .NET 9 Preview 2 below:

* [.NET Data & EF Core](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview2/efcoreanddata.md)
* [.NET MAUI](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview2/dotnetmaui.md)
* [ASP.NET Core](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview2/aspnetcore.md)

## Get started
To get started with .NET 9, [install the .NET 9 SDK](https://dotnet.microsoft.com/download/dotnet/9.0).

If you're on Windows using Visual Studio, we recommend installing the latest [Visual Studio 2022 preview](https://visualstudio.microsoft.com/vs/preview/), or get started with Visual Studio Code and the [C# Dev Kit](https://marketplace.visualstudio.com/items?itemName=ms-dotnettools.csdevkit) extension.

You can find [.NET 9 Preview 2 Installer & Binaries](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview2/9.0.0-preview.2.md) on GitHub.

## Team Announcements & Discussions
Discuss this release with the product teams on GitHub through the announcements for this release. 

* [.NET MAUI](https://github.com/dotnet/maui/discussions/21159) 
* [ASP.NET Core](https://github.com/dotnet/aspnetcore/discussions/54503)
* [Libraries & Runtime](https://github.com/dotnet/runtime/discussions/99619) 
* [Source-build](https://github.com/dotnet/source-build/discussions/4132)

## Stay up-to-date with .NET 9
You can stay up-to-date with all the features of .NET 9 with:
* [What's new in .NET 9](https://learn.microsoft.com/dotnet/core/whats-new/dotnet-9/overview)
* [What's new in ASP.NET Core](https://learn.microsoft.com/aspnet/core/release-notes/aspnetcore-9.0)
* [What's new in .NET MAUI](https://learn.microsoft.com/dotnet/maui/whats-new/dotnet-9)
* [What's new in EF Core](https://learn.microsoft.com/ef/core/what-is-new/ef-core-9.0/whatsnew)
* [Breaking Changes in .NET 9](https://learn.microsoft.com/dotnet/core/compatibility/9.0)
* [.NET 9 Releases](https://github.com/dotnet/core/blob/main/release-notes/9.0/README.md)

For .NET 9 previews we are utilizing GitHub Discussions to make announcements. For more information [read our discussion](https://github.com/dotnet/core/discussions/9131).

Stay up to date by subscribing to the [RSS feed](https://github.com/dotnet/core/discussions/categories/news.atom) for all release announcements.

**Comment by tomrus88** [comment]
So where is exact list of changes/new features/etc that come in preview 2? Old way with blogs and separate github issues for each preview before it was even out was so much better...

**Reply by hwoodiwiss** [reply]
Follow the announcement link for the team you're interested in, all of them have a link to release notes for the current preview.
Here's the [runtime](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview2/runtime.md) and [ASP.NET Core's](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview2/aspnetcore.md) for example.

**Reply by JonDouglas** [reply]
> So where is exact list of changes/new features/etc that come in preview 2? Old way with blogs and separate github issues for each preview before it was even out was so much better...

The first link in this discussion leads to release notes. Those release notes are also included in the "what's new in" documentation as well. We are still exploring this new approach, thanks for the patience.

<img width="884" alt="image" src="https://github.com/dotnet/core/assets/1288848/87646c02-4c10-46ad-88c1-a374844e4fdd">

**Reply by GerardSmit** [reply]
> The first link in this discussion leads to release notes. Those release notes are also included in the "what's new in" documentation as well. We are still exploring this new approach, thanks for the patience.

This was really confusing for me. This whole post is about ".NET 9 preview 2", but the section "What's new" is about .NET 9 in general. It would be clearer if these links were a direct link to updates of .NET 9 preview 2.

For example: to get to the runtime changes, you have to click "detailed release notes", then click on "runtime.md" because nowhere in the README.md is it clear where you can find the runtime changes. While there is a link in the README.md with "Libraries and runtime [Announcement](https://github.com/dotnet/runtime/discussions/98372)", it goes to the announcement post and not the actually changes. ~~Additional, it goes to the old announcement post (preview 1)...~~ [This has been changed](https://github.com/dotnet/core/pull/9222)

All in all, it feels messy compared to the blog posts we had in devblogs.microsoft.com. Hopefully in preview 3 it'll get better. 😅

**Reply by JonDouglas** [reply]
@GerardSmit Let me work with our documentation team to see if there's a way to provide sections in the what's new docs specifically for each preview. i.e. what's new in p1, p2, etc.

The current release notes tries to do that in detail, but it sounds like it is confusing. We will work to make it right.

**Reply by TFlippy** [reply]
I usually check the API diffs for new features/changes - would be nice to see them being posted earlier, though (since it usually takes around 1-2 weeks for them to show up).

**Reply by jamesmontemagno** [reply]
I have made adjustments to the format to make it much more streamlined.

**Comment by lofcz** [comment]
Most of the links either lead back here or to Preview 1 announcements. [What's new in .NET 9](https://learn.microsoft.com/dotnet/core/whats-new/dotnet-9/overview) is aggregated so checking out what's new in Preview 2 requires one to see the git history of the file. The old system was way more convenient.

For future readers: https://github.com/dotnet/docs/commit/e1fa64000688c3bae703f7b2d9dfbd2bd3489be4

**Reply by JonDouglas** [reply]
Would it be easier if these what's new docs had sections for each preview or similar? The current diffs can be found in the release notes: 

https://github.com/dotnet/core/tree/main/release-notes/9.0/preview

**Reply by DavidZidar** [reply]
@JonDouglas None of the files that actually contain the relevant information (sdk.md, runtime.md and so on) are actually linked anywhere in the markdown documents themselves which is confusing.

**Reply by JonDouglas** [reply]
I hear you. I'll bring this up to the team. It is a bit confusing. Thanks for the feedback.

**Comment by wthks** [comment]
this is so badly done, there are so many links and then you don't know exactly what has been changed or added because we are redirected to the documentation page.

the devblog was much easier and better to understand.

**Comment by breadnone** [comment]
The blog way was "OK", this is just heck too confusing

**Comment by JonDouglas** [comment]
Thank you all for your feedback on this process and being patient with our team as we're experimenting with different ways of communicating. We will re-assess in the meantime and bring more clarity by the next preview. For now, changes can be found in both the release notes and what's new pages.

I will make notes to ask our product and documentation teams to provide clear "diffs" of what has changed in each preview beyond the current structure of release notes which it sounds is too many clicks and not as centralized like the blogs were.

If there are certain things you would like to see from us as the community, please let us know so we can bring it up as well. Feedback as simple as "go back to the way it was" or how we can integrate some of these new concepts better are all fair game. We are listening and will adjust accordingly. 

Thanks for your patience.

**Reply by UrbanSowaLabs** [reply]
I don't have enough time for a deep dive with clicking links that lead me away from the main announcement.

The old format was great because **_I got to see all new features for that preview in one place, preferably with a code example._**

I would prefer the announcement centralized, so I can read it without clicking on links, or, at most **_one link deep._**

Main announcement:
- .NET 9
- C# 13
- ASP.NET Core
- EF Core
- MAUI

For that specific preview #, not a link to a general description for the whole .NET 9!

**Reply by wthks** [reply]
you don't even have to change anything on github, just keep publishing .net as it was on the devblog page, put links to the github there and problem solved.

it's just too confusing for some people, you have to search through all the documentation to find out what's new or changed.

the .net monthly updates are still published on the devblog, why not everything else...

**Reply by swythan** [reply]
If you could try reading the release announcement in the GitHub app on a phone and try following all the links it might be helpful. In a perfect world I'd like it if any links I have to click through to read the details of _this specific preview release_ should keep me in the app. Trying to read this announcement today kept switching me into the browser and I wasn't convinced it was actually sending me to the right pages. 

That said it would be easier not to have to click through links at all.

**Comment by kingcomxu** [comment]
so, is .net dead?

**Comment by loraderon** [comment]
Obviously (based on the feedback here), the most requested links are the Team Announcements and their related release notes.

There should be no other links above these except in the general short introduction.

Also, the note that "Announcement links will be populated throughout the release day" is very confusing. I think that this general announcement should not be posted until those announcements are already published. 

If **Team Announcements** are the first header and they include direct links to the respective release notes it would be much clearer.

Example of how this post could start:

---

We are pleased to announce the release of the second preview of .NET 9, along with comprehensive release notes that highlight significant enhancements from the initial preview. Additionally, we have previously outlined [Our Vision for .NET 9](https://aka.ms/dotnet/9/vision), which describes some of the broad areas we intend to deliver on by the upcoming November 2024 release.

# Team Announcements for Preview 2
* .NET MAUI: [Announcement & discussions](https://github.com/dotnet/maui/discussions/21159) - [Release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview2/dotnetmaui.md)
* ASP.NET Core: [Announcement & discussions](https://github.com/dotnet/aspnetcore/discussions/54503) - [Release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview2/aspnetcore.md)
* Libraries and Runtime: [Announcement & discussions](https://github.com/dotnet/runtime/discussions/99619) - [Runtime Release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview2/runtime.md) & [SDK Release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview2/sdk.md)

You can find [detailed release notes about the new features in .NET 9 Preview 2](https://github.com/dotnet/core/tree/main/release-notes/9.0/preview/preview2) and up-to-date [What's new in .NET 9](https://learn.microsoft.com/dotnet/core/whats-new/dotnet-9/overview) and [Breaking Changes in .NET 9](https://learn.microsoft.com/dotnet/core/compatibility/9.0) in .NET docs.

The latest .NET 9 release is always available at dotnet.microsoft.com and .NET 9 Releases.

*etc etc*

**Reply by jamesmontemagno** [reply]
Thanks for the feedback here and I agree, going to make this adjustment ;)

**Comment by WeihanLi** [comment]
The new `AllowOutOfOrderMetadataProperties` config in `System.Text.Json` is not included in the release notes

- https://github.com/dotnet/runtime/pull/97474
- https://github.com/dotnet/core/pull/9220/files#r1523056654

**Comment by JohnLGalt** [comment]
BTW, there are incorrect dates at 

https://github.com/dotnet/core/blob/main/release-notes/9.0/README.md#releases

Years show 2023 instead of 2024.

FYI

**Reply by jamesmontemagno** [reply]
Thanks! Sent PR, also feel free to submit a PR if you see anything else.
https://github.com/dotnet/core/pull/9227

**Comment by masonwheeler** [comment]
I don't see anything in here about assembly saving in `System.Reflection.Emit`.  Has there been any progress since Preview 1?

**Reply by jamesmontemagno** [reply]
Take a look at Preview 3 where some additional work has been done around `System.Reflection.Emit` https://github.com/dotnet/core/discussions/9271

**Comment by davepcallan** [comment]
There might be a PR pending perhaps but it looks like release notes just links back to itself?
 
![image](https://github.com/dotnet/core/assets/106764096/c1e32dcd-7542-4193-bba8-9bf7a14604fa)

Other than looking at the commits for certain tags will there still be a way to know what's coming in a preview before it officially drops? I loved this kind of format -> https://github.com/dotnet/core/issues/8133, and it helped me decide to pull down the daily's or not. 

I get issues isn't the correct format here but some page dedicated to what's new specifically in preview X as features are merged (as opposed to released) would be amazing for the enthusiast.

**Reply by jamesmontemagno** [reply]
Made adjustments here to this post and also the main readme landing. Reached out to you on X as well to further the discussion.

**Comment by MichaelPeter** [comment]
Was also wondering where the post on the .Net Blog where? So this is the new way of announcing previews, no more .net blog posts?

Can you resume the posts on the .net blog which go directly in my RSS feed in Outlook?

These posts keep me hyped and informed and I think are a perfect Marketing tool.

Same for the .NET Blog, without the Preview posts the blog becomes pointless

**Reply by jamesmontemagno** [reply]
For .NET 9 previews we are utilizing GitHub Discussions to make announcements. For more information https://github.com/dotnet/core/discussions/9131.

**Comment by HakamFostok** [comment]
In each opportunity, blog, youtube video, or whatever Microsoft does, at the end they ask for Feedback, nevertheless, when the community screams with feedback it seems that MS is not so responsive. The whole page scream for the return to the Devblogs way, that was is really much more better. I am not sure what is the perceived (or imagined) value MS saw in relinquishing the Devblogs way. I hope this will be undone as soon as possible.
I am really progressive person by nature, but that does not mean every change is good.

**Reply by masonwheeler** [reply]
> I am really progressive person by nature, but that does not mean every change is good.

"Everybody is conservative about what they know best." -- Robert Conquest's first law.

**Reply by HakamFostok** [reply]
@masonwheeler Thank you very much, I really liked the quote and I learned a new thing from you, really appreciate it, this quote will stay with me for a very long time ❤️, the best reply I ever got ❤️

**Reply by jamesmontemagno** [reply]
Based on feedback and working directly with folks on this thread we vastly updated the post from it's original form. Many of the comments were around being able to more easily find links directly to the release notes, high level overview summary, and links for RSS feed. These have all been incorporated. I even got on calls with folks to better understand how we can improve this. This new format is reflected here in the updated post and also in the Preview 3 post, on the main READMEs in the repo, and more.

We are listening as it comes to feedback around posting each release on the .NET Blog. Like any good experiment you have to give it time to actually attempt to work to gather feedback, iterate, and make decisions going forward. So trust me, myself and the team are paying very close attention to every single comment and making tweaks and adjustments as we go.
